lib/wanda/execution/evaluation.ex

```diff
@@ -66,39 +66,59 @@ defmodule Wanda.Execution.Evaluation do
           }
 
         {agent_id, facts} ->
-          %AgentCheckResult{agent_id: agent_id}
-          |> add_agent_expectation_result(facts, expectations)
+          add_agent_check_result_or_error(agent_id, facts, expectations)
           |> add_facts(facts, check_id)
       end)
 
     %CheckResult{check_result | agents_check_results: agents_results}
   end
 
-  defp add_agent_expectation_result(
-         %AgentCheckResult{} = agent_check_result,
-         facts,
-         expectations
-       ) do
-    %AgentCheckResult{
-      agent_check_result
-      | expectation_evaluations: Enum.map(expectations, &eval_expectation(&1, facts))
-    }
+  defp add_agent_check_result_or_error(agent_id, facts, expectations) do
+    if has_some_fact_gathering_error?(facts) do
+      %AgentCheckError{
+        agent_id: agent_id,
+        type: :fact_gathering_error,
+        message: "Fact gathering ocurred during the execution"
+      }
+    else
+      %AgentCheckResult{
+        agent_id: agent_id,
+        expectation_evaluations: Enum.map(expectations, &eval_expectation(&1, facts))
+      }
+    end
+  end
+
+  def has_some_fact_gathering_error?(facts) do
+    Enum.any?(facts, fn
+      {_, %{type: _, message: _}} -> true
+      _ -> false
+    end)
   end
 
   defp add_facts(%AgentCheckResult{} = agent_check_result, facts, check_id) do
     %AgentCheckResult{
       agent_check_result
-      | facts:
-          Enum.map(facts, fn
-            {name, %{type: type, message: message}} ->
-              %FactError{check_id: check_id, name: name, type: type, message: message}
+      | facts: build_facts(check_id, facts)
+    }
+  end
 
-            {name, value} ->
-              %Fact{check_id: check_id, name: name, value: value}
-          end)
+  defp add_facts(%AgentCheckError{} = agent_check_result, facts, check_id) do
+    %AgentCheckError{
+      agent_check_result
+      | facts: build_facts(check_id, facts)
     }
   end
 
+  defp build_facts(check_id, facts) do
+    Enum.map(facts, fn
+      {name, %{type: type, message: message}} ->
```

Comment:
I think the map should be populated with `{:error, fact_error}` or `{:ok, fact}` this way you can easily match without having to match the shape of the map that is a bit unclear

Reply:
the same goes for the `:timeout` condition

Reply:
Does it make sense to include an `ok, error` distinction if they are already different struct types?
It wouldn't be better simply adding the whole data struct so we can map them by the type?
Let me send a draft commit to show what I mean hehe

Reply:
@fabriziosestito Here the proposed change: https://github.com/trento-project/wanda/pull/30/commits/af69d60899ff6c643c109ae1156c9eb2a6564b70

At the end, I have changed how the fact gathering map is composed (that was pretty much attached on how Abacus works). No, the maps stores a list of Fact/FactError so we don't need to transform them back again. 

I didn't include any change in the `AgentCheckResult/AgentCheckError` usage in the `:timeout` scenario. I was feeling that the change was going to far away from the initial purpose of the PR. And in fact, the current logic doesn't look that bad, as we let the evaluation code handle all the final data creation

Reply:
no worries, if we feel to touch the timeout logic we will in the future.
i like this logic better, it seems way more readable. big LGTM